### ISSUE TITLE
voxel Probe with a stack of multiple frames

### ISSUE THREAD

**Issue by SevenSinS02** [FNNDSC/ami#90]
Hi @NicolasRannou ,

I tried the voxelProbe example with the viewers upload example. At first, i got an error "bounding sphere of null is undefined" whenever i increased the index. So I passed the mesh on each  index change. This worked and I can see the voxel-mesh but without the dom, that shows the voxel measurements.
Also, when i click, a voxel-mesh gets created however the same is visible on every frame after that index. However, any frame before that index does not display that voxel-mesh. How do I resolve this?

Thanks.

**Comment by NicolasRannou** [comment]
Hi @SevenSinS02  - do you have a code snippet somewhere?

The target mesh should be the "stackHelper.slice" and that might just work fine.

The DOM stuff should appear though, that is strange.

Finally the "voxel-mesh" behavior you describe makes sense, basically the "2D renderer" such as the one in the upload example is actually a 3D scene with an orthographic projection. Adding voxels adds it in the "3D" space so if it is "after" the slice we are looking at they will not show up. If it is "before" the slice it will appear. To show/hide it while scrolling through the stack, we basically have to check if the current slice intersects one of the voxels displayed and update the visibility accordingly.

To do so, the easiest way it go compare the stack orientation/index to the voxels in the `WidgetsVoxelProbe._voxels[]._voxel.dataCoordinates`.

Please let me know if you have more questions, would definitely be good to add that in the viewers_upload demo!

**Comment by SevenSinS02** [comment]
Hi @NicolasRannou ,

This is what I tried before. 

**viewers_upload.js**

function hookCallbacks(stackHelper) {
...
 let stackIndex=stackFolder.add(stackHelper, 'index', 0, stack.dimensionsIJK.z - 1).step(1).listen();
    stackIndex.onChange(function(value){
        console.log("stackHelper",stackHelper);
        probe._targetMesh=stackHelper.slice.mesh;// do i have to add probe to the scene again? or is this    enough?
    });
...
}

function handleSeries(seriesContainer){

...
 probe = new WidgetsVoxelProbe(stack,
                                  stackHelper.slice.mesh,
                                  controls,
                                  camera,
                                  threeD);
    probe._current._showVoxel =  true;
    probe._current._showDomSVG =  true;
    probe._current._showDomMeasurements =  true;
    scene.add(probe);
}
...
}

I will now try what you have suggested. If it works out well I'll put up the code here for others to reference as well.
Thanks!

**Comment by NicolasRannou** [comment]
That should be enough:
```
probe._targetMesh=stackHelper.slice.mesh;
```

How is `threeD` defined there?
```
probe = new WidgetsVoxelProbe(stack,
stackHelper.slice.mesh,
controls,
camera,
threeD);
```

**Comment by NicolasRannou** [comment]
Also about the DOM, are the element actually created? Could it be a CSS issue (style not defined?)
https://github.com/FNNDSC/ami/blob/dev/examples/widget_voxelProbe/widget_voxelProbe.css

**Comment by SevenSinS02** [comment]
Omg! It actually was a css issue. I can't believe I missed that. Apologies for wasting your time on that.

The threeD is left untouched from your example. It's the div container that has the renderer.

threeD = document.getElementById('r3d');
  renderer = new THREE.WebGLRenderer({
    antialias: true
  });
  renderer.setSize(threeD.clientWidth, threeD.clientHeight);
  renderer.setClearColor(0x212121, 1);

  threeD.appendChild(renderer.domElement);

**Comment by NicolasRannou** [comment]
No worries! Please let me know if you get a nice working demo, that could be nice to integrate in the `viewers_upload` example!

**Comment by SevenSinS02** [comment]
Hi @NicolasRannou ,

I just wanted to say that it works now. I tried as you suggested and set probe._voxels[i]._mesh.visible=true/false; based on the condition. Shouldn't
 setting probe._voxels[i].showVoxel=true/false would be enough? The code is a bit of a mess. Let me know if I could make it better. Thanks! If you are satisfied with the code I could probably send a pull request .

   function checkVoxel(stackHelper){

//only tried it for axial

       let o=stackHelper.orientation,
            ind=stackHelper.index;   
      
      for(let i=0;i<probe._voxels.length;i++){
         
         if(o==2){
        if(probe._voxels[i]._voxel.dataCoordinates['y']==ind){
             probe._voxels[i]._mesh.visible=true;
        }
        else probe._voxels[i]._mesh.visible=false;
         } else if(o==1){
            if(probe._voxels[i]._voxel.dataCoordinates['x']==ind){
               probe._voxels[i]._mesh.visible=true;
            }
            else probe._voxels[i]._mesh.visible=false;
         }
         else if(o==0){
           if(probe._voxels[i]._voxel.dataCoordinates['z']==ind){
              probe._voxels[i]._mesh.visible=true;
              probe._voxels[i]._showDomSVG =true;
               probe._voxels[i]._showDomMeasurements =true;
            }else {
              //probe._voxels[i]._showVoxel=false;
                probe._voxels[i]._mesh.visible=false;
                 probe._voxels[i]._mesh.visible=false;
              probe._voxels[i]._showDomSVG =false;
               probe._voxels[i]._showDomMeasurements =false;
          }
         }
         else throw "orientation not set";
       }
    }

function hookCallbacks(stackHelper) {
    let stack = stackHelper._stack;
controls.addEventListener('OnScroll', function(e) {
      if (e.delta > 0) {
        if (stackHelper.index >= stack.dimensionsIJK.z - 1) {
          return false;
        }
        stackHelper.index += 1;
       probe._targetMesh=stackHelper.slice.mesh;
      checkVoxel(stackHelper);
      } else {
        if (stackHelper.index <= 0) {
          return false;
        }
        stackHelper.index -= 1;
       probe._targetMesh=stackHelper.slice.mesh;
       checkVoxel(stackHelper);
      }
    });
..
}

**Comment by NicolasRannou** [comment]
Yes you are right, the following should be enough:
```
  probe._voxels[i].showVoxel=false;
  probe._voxels[i].showDomSVG =false;
  probe._voxels[i].showDomMeasurements =false;
```

**Comment by NicolasRannou** [comment]
@SevenSinS02 instead of the huge else/if statement you should try the following: (moreless)
```
let slice = stackHelper.slice;
let planePosition = slice.planePosition;
let planeDirection = slice.planeDirection;
let toAABB = new THREE.Matrix4();
let voxelHalfDimensions = new THREE.Vector3(0.5, 0.5, 0.5);
let show = false;

// loop through all the voxels
for(let i=0;i<probe._voxels.length;i++){
      let voxelCenter = probe._voxels[i]._voxel.dataCoordinates;
      let aabb = {
        voxelHalfDimensions,
        voxelCenter,
        toAABB,
      };

      let plane = {
        planePosition,
        planeDirection,
      };

      // BOOM!
      let intersections = coreIntersections.aabbPlane(aabb, plane);

      // true if >= 3 intersection
      // false if less
      show = (intersections.length >= 3);

     probe._voxels[i].showVoxel=show;
     probe._voxels[i].showDomSVG =show;
     probe._voxels[i].showDomMeasurements =show;
}
```

**Comment by NicolasRannou** [comment]
https://github.com/FNNDSC/ami/blob/master/src/core/core.intersections.js

**Comment by SevenSinS02** [comment]
Hi @NicolasRannou,

I didn't even think of that at all. This very much makes the code much cleaner. I will post an example with this tomorrow.
  Thanks!

**Comment by SevenSinS02** [comment]
Hi @NicolasRannou ,

Apologies for responding so late. I've been busy with other stuff and wasn't able to post the example.
I had to make some changes to the code in widgets.voxelProbe.js. I had to remove the lines below as they always made sure the voxels and the doms were seen.

Line 304 - 307.

      // show hide mesh
      this._voxels[i].showVoxel = this._showVoxel;
      // show/hide dom stuff
      this._voxels[i].showDomSVG = this._showDomSVG;
      this._voxels[i].showDomMeasurements = this._showDomMeasurements;

If this is fine with you, then I'll send a pull request to you.
